src/main/java/org/elasticsearch/example/realm/CustomCachingRealm.java

```diff
@@ -64,16 +65,26 @@ public User authenticate(AuthenticationToken authenticationToken) {
         if (userHolder == null || userHolder.password == null) {
             User user = super.authenticate(token);
             if (user != null) {
-                userHolder = new UserHolder(token.credentials().copyChars(), user);
+                userHolder = new UserHolder(token.credentials().getChars(), user);
```

Comment:
This needs to clone the characters.
The `getChars` method returns the underlying array inside the `SecureString`. When the `SecureString` is closed that array is filled with `'\0'`, but this method will have stored that array in the cache.